### PR TITLE
Add `into_inner` and `get_mut` to `Mutex` and `RwLock`

### LINKED ISSUE #28968: Tracking Issue for Mutex and RWLock unique accessors
After #29031, this is now the tracking issue for the stabilization of the following features:
 * [ ] mutex_into_inner
 * [ ] mutex_get_mut
 * [ ] rwlock_into_inner
 * [ ] rwlock_get_mut

## Original report

As far as I can tell, it's sound to straight-up-unwrap a Mutex if you have it by value. This would enable one to do:

```rust
let data = Mutex::new(vec![])
crossbeam::scope(|scope| {
  for input in inputs {
    let data = &data;
    scope.spawn(move || {
      let output = expensive_computation(input);
      data.lock().unwrap().push(output);
    }
  }
});

for expensive in data.into_inner() {
  non_thread_safe(expensive);
}
```

Which today requires mutexing an `&mut`:

```rust
let mut data = vec![];
{
  let data_mutex = Mutex::new(&mut data);
  crossbeam::scope(|scope| {
    for input in inputs {
      let data = &data_mutex;
      scope.spawn(move || {
        let output = expensive_computation(input);
        data.lock().unwrap().push(output);
      }
    }
  });
}

for expensive in data.into_inner() {
  non_thread_safe(expensive);
}
```

Given Arc can be unwrapped, this would also allow the last `Arc<Mutex<T>>` in some concurrent context to be fully unwrapped to a `T`.

### PR DESCRIPTION
**Pull request by cristicbz**
The implementation for `into_inner` was a bit more complex than I had hoped for---is there any simpler, less unsafe way of getting around the fact that one can't move out of a `Drop` struct?

See #28968 and rust-lang/rfcs#1269 .

### CODE REVIEW

**Comment by cristicbz** [comment]
cc @Gankro @tomaka @alexcrichton

File: src/libstd/sync/mutex.rs [hunk 1]
```diff
@@ -243,6 +245,55 @@ impl<T: ?Sized> Mutex<T> {
...
+        // `self` so there's no need to lock the inner StaticMutex.
+        //
+        // To get the inner value, we'd like to call `data.into_inner()`,
+        // but because `Mutex` impl-s `Drop`, we can't move out of it, so
+        // we'll have to destructure it manually instead.
+        unsafe {
+            // The following code is like `let Mutex { inner, data } = self`.
+            let mut inner = mem::uninitialized::<Box<StaticMutex>>();
+            let mut data = mem::uninitialized::<UnsafeCell<T>>();
+            ptr::copy_nonoverlapping(&self.inner as *const Box<StaticMutex>,
```
Comment (Gankro) [inline src/libstd/sync/mutex.rs]: Just use ptr::read

**Comment by Gankro** [comment]
Other than the ptr::read stuff, this looks great! Really thorough. I'm not really qualified to review this kind of code, though.

File: src/libstd/sync/mutex.rs (hunk 1 above)
Comment (Gankro) [inline src/libstd/sync/mutex.rs]: Also note that &T coerces to *T

Comment (cristicbz) [inline src/libstd/sync/mutex.rs]: Done.

**Comment by cristicbz** [comment]
Thanks @Gankro! I changed to `ptr::read`, I'd forgotten about that one.

**Comment by cristicbz** [comment]
Local `make check` just finished and passed.

**Comment by cristicbz** [comment]
@aturon mentioned in a different bug he is going to be unavailable. r? @alexcrichton

**Comment by huonw** [comment]
This shouldn't close #28968 if it is using it as a tracking issue.

**Comment by alexcrichton** [comment]
The libs team talked about this today, and the conclusion is that this is good to go. The changes we'd like to see, however, are:

* Could you add these methods to `RwLock` as well? They'd basically all be doing the same thing.
* It may be good to get some exhaustive matching here to ensure that all struct fields are moved out of when `into_inner` is called. For example something like:

```rust
let (a, b) = {
    let Mutex { ref a, ref b } = self; // note exhaustive match
    (ptr::read(a), ptr::read(b))
};
mem::forget(self);
```

* Could you add a comment in the destructor as well indicating that `into_inner` needs to be updated with any changes as well?

**Comment by cristicbz** [comment]
@huonw Good point, I edited the description to not close the issue.

@alexcrichton That all makes sense, I'll do all of that when I get back from work, thanks!

**Comment by cristicbz** [comment]
@alexcrichton I implemented everything you asked for, please take a look.

Thanks!

File: src/libstd/sync/mutex.rs [hunk 2]
```diff
@@ -243,6 +245,52 @@ impl<T: ?Sized> Mutex<T> {
...
         self.inner.poison.get()
     }
+
+    /// Consumes this mutex, returning the underlying data.
+    ///
+    /// # Failure
+    ///
+    /// If another user of this mutex panicked while holding the mutex, then
+    /// this call will return an error instead.
+    #[inline]
```
Comment (alexcrichton) [inline src/libstd/sync/mutex.rs]: Ah as one final piece, could these `#[inline]` annotations be left out? These are pretty hefty functions and the inline shouldn't be needed anyway as a result of the generic parameter on `Mutex`

**Comment by cristicbz** [comment]
@alexcrichton Done and squashed. Don't feel strongly, but I had added `#[inline]` before because (a) they're not actually that big (especially get_mut) and (b) consistency with the other methods on the types.  That said I don't expect these methods to be called in a tight loop (event `get_mut`), so this is probably for the better.

The fact that it has a type param doesn't quite imply `#[inline]`, right? It's still a hint for LLVM---I remember when we inlined all the slice methods (also type param'd) and it made a massive impact on perf.

**Comment by alexcrichton** [comment]
@bors: r+ 90ccefd16e1eeec1a616aa3ba9aa710d954231ef

Thanks!

Yeah `#[inline]` is an extra hint on top of monomorphizing to LLVM but it rarely has a massive impact on perf for generic functions. The real impact it has is when a function is otherwise not translated across crates (e.g. a concrete function like `str::len`).

**Pull request merged**